vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -37,102 +37,69 @@
  * NOTE: Currently, only SMALLEST is supported.
  *
  * See e_representative_entry_method */
-#define REPRESENTATIVE_ENTRY_METHOD util::SMALLEST
-// #define FILL_LIMIT 30
+#define REPRESENTATIVE_ENTRY_METHOD SMALLEST
 
 #define CONNECTION_BOX_LOOKAHEAD_MAP_PRINT_COST_MAPS
 
 // Sample based an NxN grid of starting segments, where N = SAMPLE_GRID_SIZE
 static constexpr int SAMPLE_GRID_SIZE = 3;
 
+// Don't continue storing a path after hitting a lower-or-same cost entry.
+static constexpr bool BREAK_ON_MISS = false;
+
 // Distance penalties filling are calculated based on available samples, but can be adjusted with this factor.
 static constexpr float PENALTY_FACTOR = 1.f;
+static constexpr float PENALTY_MIN = 1e-12f;
+
+static constexpr int MIN_PATH_COUNT = 1000;
 
 // a sample point for a segment type, contains all segments at the VPR location
 struct SamplePoint {
-    uint64_t order; // used to order sample points
     vtr::Point<int> location;
-    std::vector<ssize_t> samples;
-    SamplePoint()
-        : location(0, 0) {}
+    std::vector<ssize_t> nodes;
 };
 
-// a grid of sample points
-struct SampleGrid {
-    SamplePoint point[SAMPLE_GRID_SIZE][SAMPLE_GRID_SIZE];
-};
-
-// implements a simple cache of key(K)/value(V) pairs of N entries
-template<class K, class V, int N>
-class SimpleCache {
-  public:
-    SimpleCache()
-        : pos(0)
-        , hits(0)
-        , misses(0) {}
-
-    // O(N) lookup
-    bool get(K key, V* value) {
-        for (int i = 0; i < N; i++) {
-            auto& k = keys[i];
-            if (k == key) {
-                auto& v = values[i];
-#if defined(CONNECTION_BOX_LOOKAHEAD_PUSH_BACK_HITS)
-                // preserve the found key by pushing it back
-                int last = (pos + N - 1) % N;
-                std::swap(k, keys[last]);
-                std::swap(v, values[last]);
-#endif
-                *value = v;
-                hits++;
-                return true;
-            }
-        }
-        misses++;
-        return false;
-    }
-
-    // O(1) insertion (overwriting an older entry)
-    void insert(K key, V val) {
-        keys[pos] = key;
-        values[pos] = val;
-        pos = (pos + 1) % N;
-    }
-
-    // ratio of successful lookups
-    float hit_ratio() {
-        return hits ? static_cast<float>(hits) / static_cast<float>(hits + misses) : 0.f;
-    }
-
-    // ratio of unsuccessful lookups
-    float miss_ratio() {
-        return misses ? static_cast<float>(misses) / static_cast<float>(hits + misses) : 0.f;
-    }
-
-  private:
-    std::array<K, N> keys; // keep keys together for faster scanning
-    std::array<V, N> values;
-    size_t pos;
-    uint64_t hits;
-    uint64_t misses;
+struct SampleRegion {
+    int segment_type;
+    vtr::Point<int> grid_location;
+    std::vector<SamplePoint> points;
+    uint64_t order; // for sorting
```

Comment:
Is `order` a good variable name?  The comment actually confuses me more than it explains it's purpose.  Maybe expand the comment and/or come up with a clearer name?

Reply:
Comment added.